File: src/metrics/emergence/add_observations.py
class TransferEntropy:
    def __init__(self, start_observation_time, dest_embedding_delay, k, base, source_embedding_delay,
                 source_history_embed_length, delay, source_next_past_count, source_past_count,
                 next_past_count, past_count, next_count, max_shifted_value, max_shifted_source_value):
        self.start_observation_time = start_observation_time
        self.dest_embedding_delay = dest_embedding_delay
        self.k = k
        self.base = base
        self.source_embedding_delay = source_embedding_delay
        self.source_history_embed_length = source_history_embed_length
        self.delay = delay
        self.source_next_past_count = source_next_past_count
        self.source_past_count = source_past_count
        self.next_past_count = next_past_count
        self.past_count = past_count
        self.next_count = next_count
        self.max_shifted_value = max_shifted_value
        self.max_shifted_source_value = max_shifted_source_value
        self.observations = 0

    def add_observations(self, source, dest, start_time, end_time):
        if (end_time - start_time) - self.start_observation_time + 1 <= 0:
            # No observations to add
            return
        if end_time >= len(dest) or end_time >= len(source):
            raise IndexError(f"endTime ({end_time}) must be <= length of input arrays (dest: {len(dest)}, source: {len(source)})")
        
        # Increment the count of observations
        self.observations += (end_time - start_time) - self.start_observation_time + 1

        # Initialize and store the current previous values;
        #  one for each phase of the embedding delay.
        # First for the destination:
        past_val = np.zeros(self.dest_embedding_delay, dtype=int)
        for d in range(self.dest_embedding_delay):
            # Compute the current previous values for
            #  phase d of the embedding delay, but leave
            #  out the most recent value (we'll add those in
            #  in the main loop)
            past_val[d] = 0
            for p in range(self.k - 1):
                past_val[d] += dest[start_time + self.start_observation_time + d - 1
                                    - (self.k - 1) * self.dest_embedding_delay
                                    + p * self.dest_embedding_delay]
                past_val[d] *= self.base

        # Next for the source:
        source_past_val = np.zeros(self.source_embedding_delay, dtype=int)
        for d in range(self.source_embedding_delay):
            # Compute the current previous values for
            #  phase d of the embedding delay, but leave
            #  out the most recent value (we'll add those in
            #  in the main loop)
            source_past_val[d] = 0
            for p in range(self.source_history_embed_length - 1):
                source_past_val[d] += source[start_time + self.start_observation_time + d - self.delay
                                             - (self.source_history_embed_length - 1) * self.source_embedding_delay
                                             + p * self.source_embedding_delay]
                source_past_val[d] *= self.base

        # 1. Count the tuples observed
        dest_val = 0
        dest_embedding_phase = 0
        source_embedding_phase = 0

        for r in range(start_time + self.start_observation_time, end_time + 1):
            # First update the embedding values for the current
            #  phases of the embeddings:
            if self.k > 0:
                past_val[dest_embedding_phase] += dest[r - 1]
            source_past_val[source_embedding_phase] += source[r - self.delay]
            # Add to the count for this particular transition:
            dest_val = dest[r]
            this_past_val = past_val[dest_embedding_phase]
            this_source_val = source_past_val[source_embedding_phase]
            self.source_next_past_count[this_source_val][dest_val][this_past_val] += 1
            self.source_past_count[this_source_val][this_past_val] += 1
            self.next_past_count[dest_val][this_past_val] += 1
            self.past_count[this_past_val] += 1
            self.next_count[dest_val] += 1

            # Now, update the combined embedding values and phases,
            #  for this phase we back out the oldest value which we'll no longer need:
            if self.k > 0:
                past_val[dest_embedding_phase] -= self.max_shifted_value[dest[r - 1 - (self.k - 1) * self.dest_embedding_delay]]
                past_val[dest_embedding_phase] *= self.base  # and shift the others up
            source_past_val[source_embedding_phase] -= \
                self.max_shifted_source_value[
                    source[r - self.delay - (self.source_history_embed_length - 1) * self.source_embedding_delay]]
            source_past_val[source_embedding_phase] *= self.base  # and shift the others up

            # then update the phase
            dest_embedding_phase = (dest_embedding_phase + 1) % self.dest_embedding_delay
            source_embedding_phase = (source_embedding_phase + 1) % self.source_embedding_delay

File: src/metrics/emergence/add_observations.java
	public void addObservations(int[] source, int[] dest) {
		addObservations(source, dest, 0, dest.length-1);
	}

/**
 	 * Add observations for a single source-destination pair
 	 *  to our estimates of the pdfs.
	 * Start and end time are the (inclusive) indices within which to add the observations.
	 * The start time is from the earliest of the k historical values of the destination (inclusive),
	 *  the end time is the last destination time point to add in.
	 *  
	 * @param source source time-series
	 * @param dest destination time-series. 
	 *  Must be same length as source
	 * @param startTime earliest time that we may extract embedded history from
	 * @param endTime last destination (next) time point to add in
	 * 
	 */
	public void addObservations(int[] source, int[] dest, int startTime, int endTime) {
		if ((endTime - startTime) - startObservationTime + 1 <= 0) {
			// No observations to add
			return;
		}
		if ((endTime >= dest.length) || (endTime >= source.length)) {
			throw new ArrayIndexOutOfBoundsException(
					String.format("endTime (%d) must be <= length of input arrays (dest: %d, source: %d)",
							endTime, dest.length, source.length));
		}
		// increment the count of observations:
		observations += (endTime - startTime) - startObservationTime + 1; 
		
		// Initialise and store the current previous values;
		//  one for each phase of the embedding delay.
		// First for the destination:
		int[] pastVal = new int[destEmbeddingDelay];
		for (int d = 0; d < destEmbeddingDelay; d++) {
			// Compute the current previous values for
			//  phase d of the embedding delay, but leave
			//  out the most recent value (we'll add those in
			//  in the main loop)
			pastVal[d] = 0;
			for (int p = 0; p < k-1; p++) {
				pastVal[d] += dest[startTime + startObservationTime + d - 1
				                   - (k-1)*destEmbeddingDelay
				                   + p*destEmbeddingDelay];
				pastVal[d] *= base;
			}
		}
		// Next for the source:
		int[] sourcePastVal = new int[sourceEmbeddingDelay];
		for (int d = 0; d < sourceEmbeddingDelay; d++) {
			// Compute the current previous values for
			//  phase d of the embedding delay, but leave
			//  out the most recent value (we'll add those in
			//  in the main loop)
			sourcePastVal[d] = 0;
			for (int p = 0; p < sourceHistoryEmbedLength - 1; p++) {
				sourcePastVal[d] += source[startTime + startObservationTime + d - delay
				                    - (sourceHistoryEmbedLength-1)*sourceEmbeddingDelay
				                    + p*sourceEmbeddingDelay];
				sourcePastVal[d] *= base;
			}
		}
		
		// 1. Count the tuples observed
		int destVal, destEmbeddingPhase = 0, sourceEmbeddingPhase = 0;
		for (int r = startTime + startObservationTime; r <= endTime; r++) {
			// First update the embedding values for the current
			//  phases of the embeddings:
			if (k > 0) {
				pastVal[destEmbeddingPhase] += dest[r-1];
			}
			sourcePastVal[sourceEmbeddingPhase] += source[r-delay];
			// Add to the count for this particular transition:
			// (cell's assigned as above)
			destVal = dest[r];
			int thisPastVal = pastVal[destEmbeddingPhase];
			int thisSourceVal = sourcePastVal[sourceEmbeddingPhase];
			sourceNextPastCount[thisSourceVal][destVal][thisPastVal]++;
			sourcePastCount[thisSourceVal][thisPastVal]++;
			nextPastCount[destVal][thisPastVal]++;
			pastCount[thisPastVal]++;
			nextCount[destVal]++;
			// Now, update the combined embedding values and phases,
			//  for this phase we back out the oldest value which we'll no longer need:
			if (k > 0) {
				pastVal[destEmbeddingPhase] -= maxShiftedValue[dest[r-1-(k-1)*destEmbeddingDelay]];
				pastVal[destEmbeddingPhase] *= base; // and shift the others up
			}
			sourcePastVal[sourceEmbeddingPhase] -=
					maxShiftedSourceValue[
					    source[r-delay-(sourceHistoryEmbedLength-1)*sourceEmbeddingDelay]];
			sourcePastVal[sourceEmbeddingPhase] *= base; // and shift the others up
			// then update the phase
			destEmbeddingPhase = (destEmbeddingPhase + 1) % destEmbeddingDelay; 
			sourceEmbeddingPhase = (sourceEmbeddingPhase + 1) % sourceEmbeddingDelay; 
		}
	}